R/bm-tpc-h.R

```diff
@@ -218,20 +232,397 @@ tpc_h_queries[[1]] <- function(input_func) {
+    ocnr, by=c("l_orderkey" = "o_orderkey")) %>%
+    select(l_partkey, l_suppkey, l_extendedprice, l_discount, o_orderdate)
+
+  locnrp <- inner_join(locnr,
+                       input_func("part") %>%
+                         select(p_partkey, p_type) %>%
+                         filter(p_type == "ECONOMY ANODIZED STEEL") %>%
+                         select(p_partkey),
+                       by = c("l_partkey" = "p_partkey")) %>%
+    select(l_suppkey, l_extendedprice, l_discount, o_orderdate)
+
+  locnrps <- inner_join(locnrp,
+                        input_func("supplier") %>%
+                          select(s_suppkey, s_nationkey),
+                        by = c("l_suppkey" = "s_suppkey")) %>%
+    select(l_extendedprice, l_discount, o_orderdate, s_nationkey)
+
+  all <- inner_join(locnrps,
+                    input_func("nation") %>%
+                      select(n2_nationkey = n_nationkey, n2_name = n_name),
+                    by = c("s_nationkey" = "n2_nationkey")) %>%
+    select(l_extendedprice, l_discount, o_orderdate, n2_name)
+
+  aggr <- all %>%
+    mutate(
+      o_year = as.integer(strftime(o_orderdate, "%Y")),
+      volume = l_extendedprice * (1 - l_discount),
+      nation = n2_name) %>%
+    select(o_year, volume, nation) %>%
+    group_by(o_year) %>%
+    summarise(mkt_share = sum(ifelse(nation == "BRAZIL", volume, 0)) / sum(volume)) %>%
+    arrange(o_year)
+
+  aggr %>%
+    collect()
+}
+
+tpc_h_queries[[9]] <- function(input_func) {
+  p <- input_func("part") %>%
+    select(p_name, p_partkey) %>%
+    filter(grepl(".*green.*", p_name)) %>%
+    select(p_partkey)
+
+  psp <- inner_join(
+    input_func("partsupp") %>%
+      select(ps_suppkey, ps_partkey, ps_supplycost),
+    p, by = c("ps_partkey" = "p_partkey"))
+
+  sn <- inner_join(
+    input_func("supplier") %>%
+      select(s_suppkey, s_nationkey),
+    input_func("nation") %>%
+      select(n_nationkey, n_name),
+    by = c("s_nationkey" = "n_nationkey")) %>%
+    select(s_suppkey, n_name)
+
+  pspsn <- inner_join(psp, sn, by = c("ps_suppkey" = "s_suppkey"))
+
+  lpspsn <- inner_join(
+    input_func("lineitem") %>%
+      select(l_suppkey, l_partkey, l_orderkey, l_extendedprice, l_discount, l_quantity),
+    pspsn,
+    by = c("l_suppkey" = "ps_suppkey", "l_partkey" = "ps_partkey")) %>%
+    select(l_orderkey, l_extendedprice, l_discount, l_quantity, ps_supplycost, n_name)
+
+  all <- inner_join(
+    input_func("orders") %>%
+      select(o_orderkey, o_orderdate),
+    lpspsn,
+    by = c("o_orderkey"= "l_orderkey" )) %>%
+    select(l_extendedprice, l_discount, l_quantity, ps_supplycost, n_name, o_orderdate)
+
+  aggr <- all %>%
+    # kludge, should not need to collect here, but if we don't the answers we
+    # get are wrong (though they *do* complete).
+    collect() %>%
+    mutate(
+      nation = n_name,
+      # kludge, o_year = as.integer(format(o_orderdate, "%Y")),
+      o_year = as.integer(strftime(o_orderdate, "%Y")),
```

Comment:
perhaps, or why not?

```suggestion
      o_year = year(o_orderdate),
```

Reply:
The only reason is I didn't realize we had all of those functions implemented already! I need to go back and run all of these through duckdb too. We don't *need* to keep these pipelines compatible with both if we don't want to, but it's a nice sanity check to be able to run it against a second backend (especially when we go to larger scales where we don't have the full answers)

Reply:
duckdb might have `year()` or dbplyr might translate it appropriately without us needing to do anything, but that's one reason I haven't been simplifying these as I go

Reply:
got it, this is copied from the other repo

Reply:
Yeah, and even when we do starting drafting them on their own keeping them compatible is worth _a bit_ of effort.

It also shouldn't be too hard (though I've not put in the time to do it) to run these against pure-R dplyr too. It should be "just" making a new set of `input_funcs`